src/login.rs

```diff
@@ -182,20 +211,69 @@ impl Login {
 
         let self_ = imp::Login::from_instance(self);
         let visible_page = self_.content.visible_child_name().unwrap();
-        if visible_page == "phone-number-page" {
-            let encryption_key = self_.custom_encryption_key_entry.text().to_string();
-            if !encryption_key.is_empty() {
-                self.change_encryption_key();
+
+        match visible_page.as_str() {
+            "phone-number-page" => {
+                let encryption_key = self_.custom_encryption_key_entry.text().to_string();
+                if !encryption_key.is_empty() {
+                    self.change_encryption_key();
+                }
+
+                self.send_phone_number();
+            }
+            "code-page" => self.send_code(),
+            "registration-page" => {
+                if !self_.show_tos_popup.get() || self_.tos_accepted_by_user.get() {
```

Comment:
I guess you added the `tos_accepted_by_user` variable to avoid asking again to the user if he/she wants to accept the tos when the registration failed previously. I'd say to just be safe and remove the `tos_accepted_by_user` variable completely and ask again if the user wants to accept the tos. The only conceivable way to fail the registration is not setting a name, so it's not likely to happen frequently anyway.

Reply:
You are correct.